sdks/cpp/src/agones/sdk.cc

```diff
@@ -26,8 +28,33 @@ struct SDK::SDKImpl {
   std::unique_ptr<grpc::ClientContext> health_context_;
 };
 
+std::string port() {
```

Comment:
I think the use of `std::getenv()` here is good. For parsing strings to numbers you generally want to use something like `std::strtoll` rather than `sscanf`, unless you're parsing a complicated string.

However, in this case, I slightly question the need to parse and validate the port number, then just to turn it back into a string again. My suggest would be to always use the value in the env var if present, otherwise use the default. Something like:

```cc
namespace {
std::string GetPort() {
  const char* port = std::getenv("AGONES_SDK_GRPC_PORT");
  return port ? port : "5937";
}
}  // namespace
```

What do you think?

Reply:
The advantage of the current code is that if you pass an invalid value for the port (not a number or a number that isn't a valid port number to bind to) then you get both an error message and a client that attempts to connect to the default port:

```
$ docker run -e AGONES_SDK_GRPC_PORT=sathoeu gcr.io/robertbailey-agones/cpp-simple-server:0.10
C++ Game Server has started!
Getting the instance of the SDK.
Attempting to connect...
Unable to parse 'sathoeu' defined in AGONES_SDK_GRPC_PORT into an integer
```

With your suggestion, I get:

```
$ docker run -e AGONES_SDK_GRPC_PORT=stahou gcr.io/robertbailey-agones/cpp-simple-server:0.10
C++ Game Server has started!
Getting the instance of the SDK.
Attempting to connect...
Could not connect to the sidecar. Exiting!
```

Reply:
I've replaced `sscanf` with `std::strtoll` but I think that having the extra parsing (and error messages), while much more verbose than your suggestion (which also works) is a better experience for anyone trying to debug connection issues between their game and the sdkserver sidecar container. In practice, this shouldn't actually happen, so maybe it's just me overtraining on my attempts to debug this while testing modifying port numbers, and it's more code to maintain, so if you feel strongly then I'm happy to acquiesce.

Reply:
OK. I've made some comments below in case you want to keep this port validation code. Of course, I'll approve with this.

But I will try to clarify why I think this is a mistake.

1. This function isn't using the port, so I claim this is the wrong place to do any verification of the port. Does this function know *how* the port is going to be used? Is this function *sure* that the port must be a number and not, say, an alphanumeric service name (like from `/etc/services`), or something else? This function is enforcing some policy about what makes a valid port number without any knowledge of what that policy should or needs to be.

2. Validating the port number here is non-trivial. A 20-line function is more likely to have bugs than is a 2-line function.

The code that *should* validate the port number is the code that actually uses it. If an invalid port number is given, does `grpc::CreateChannels` return an error (or nullptr)? If so, that's where and how we should report the error.

In your example above that prints "Could not connect to the sidecar", can *that* error message simply include the grpc target it was trying to connect to? For example, if it said "Could not connect to the sidecar at 127.0.0.1:stahou" that would make it clear to the user what the problem is.

Reply:
Maybe it isn't the best SDK design, but the `Connect` function just returns a boolean (success or failure) and doesn't log any sort of error message (the client can if they want to, which is what we see here). The address that is being connected to is opaque to the client, and the client has no way to know what address to add to the log message. 

The reason I added the output with your suggestion is to show that (a) `grpc::CreateChannels` successfully creates a "channel" to `127.0.0.1:stahou` and (b) `channel_->WaitForConnected` attempts to connect and just times out. 

I added the extra error message because we happen to know that in our case the thing we are connecting to is a port on localhost. Since we know more information about the endpoint than the libraries, we can add validation and error messages that the libraries can't (they probably do work with an alphanumeric service name since they are designed to be more flexible). 

In practice, this should only be an issue for people trying to run things manually outside of agones, so I think adding an error message from inside the `Connect` function might be the simplest thing to do (rather than adding error messages inside of the constructor). 

Here's what the output (from the simple server) looks like after this change:

```
$ docker run --net host -e AGONES_SDK_GRPC_PORT=3344aaa  gcr.io/robertbailey-agones/cpp-simple-server:0.10
C++ Game Server has started!
Getting the instance of the SDK.
Attempting to connect...
Could not connect to the sidecar at localhost:3344aaa.
Exiting!
```